### PR TITLE
fix mcjit debuginfo

### PR DESCRIPTION
**Pull request by vtjnash**

### CODE REVIEW

File: examples/embedding.c [hunk 1]
```diff
@@ -99,6 +99,7 @@ int main()
         }
     }
 
-    jl_atexit_hook();
-    return 0;
+    int ret = 0;
+    jl_atexit_hook(ret);
```
Comment (tkelman) [inline examples/embedding.c]: also wrong in build_executable, devdocs/eval.rst, devdocs/init.rst, and devdocs/embedding.rst

**Comment by tkelman** [comment]
ftbfs on win64, llvm 3.7.0:
```
    CC src/debuginfo.o
/home/Tony/julia/src/debuginfo.cpp: In member function ‘virtual void JuliaJITEventListener::NotifyObjectEmitted(const llvm::object::ObjectFile&, const llvm::RuntimeDyld::LoadedObjectInfo&)’:
/home/Tony/julia/src/debuginfo.cpp:227:35: error: no matching function for call to ‘llvm::object::SymbolRef::getName(llvm::StringRef&) const’
             sym_iter.getName(sName);
                                   ^
/home/Tony/julia/src/debuginfo.cpp:227:35: note: candidate is:
In file included from /home/Tony/julia/usr/include/llvm/DebugInfo/DIContext.h:20:0,
                 from /home/Tony/julia/usr/include/llvm/ExecutionEngine/RuntimeDyld.h:21,
                 from /home/Tony/julia/usr/include/llvm/ExecutionEngine/ExecutionEngine.h:18,
                 from /home/Tony/julia/src/debuginfo.cpp:6:
/home/Tony/julia/usr/include/llvm/Object/ObjectFile.h:306:27: note: llvm::ErrorOr<llvm::StringRef> llvm::object::SymbolRef::getName() const
 inline ErrorOr<StringRef> SymbolRef::getName() const {
                           ^
/home/Tony/julia/usr/include/llvm/Object/ObjectFile.h:306:27: note:   candidate expects 0 arguments, 1 provided
/home/Tony/julia/src/debuginfo.cpp:288:64: error: ‘SectionAddr’ was not declared in this scope
         *(DWORD*)&UnwindData[8] = (DWORD)(catchjmp - (uint8_t*)SectionAddr); // relative location of catchjmp
                                                                ^
/home/Tony/julia/src/debuginfo.cpp:318:35: error: no matching function for call to ‘llvm::object::SymbolRef::getName(llvm::StringRef&) const’
             sym_iter.getName(sName);
                                   ^
/home/Tony/julia/src/debuginfo.cpp:318:35: note: candidate is:
In file included from /home/Tony/julia/usr/include/llvm/DebugInfo/DIContext.h:20:0,
                 from /home/Tony/julia/usr/include/llvm/ExecutionEngine/RuntimeDyld.h:21,
                 from /home/Tony/julia/usr/include/llvm/ExecutionEngine/ExecutionEngine.h:18,
                 from /home/Tony/julia/src/debuginfo.cpp:6:
/home/Tony/julia/usr/include/llvm/Object/ObjectFile.h:306:27: note: llvm::ErrorOr<llvm::StringRef> llvm::object::SymbolRef::getName() const
 inline ErrorOr<StringRef> SymbolRef::getName() const {
                           ^
/home/Tony/julia/usr/include/llvm/Object/ObjectFile.h:306:27: note:   candidate expects 0 arguments, 1 provided
/home/Tony/julia/src/debuginfo.cpp: In function ‘DWORD64 jl_getUnwindInfo(ULONG64)’:
/home/Tony/julia/src/debuginfo.cpp:1050:68: error: ‘struct ObjectInfo’ has no member named ‘SectionSize’
     if (it != objmap.end() && (intptr_t)(*it).first + (*it).second.SectionSize > dwAddr) {
                                                                    ^
Makefile:98: recipe for target 'debuginfo.o' failed
make[1]: *** [debuginfo.o] Error 1
Makefile:87: recipe for target 'julia-src-release' failed
make: *** [julia-src-release] Error 2
```

**Comment by vtjnash** [comment]
@tkelman i think i fixed the build issues noted above. can you test that backtraces look ok?

**Comment by tkelman** [comment]
It compiles libjulia now, but still segfaults right away in bootstrap:
```
$ gdb -q --args /home/Tony/julia/usr/bin/julia-debug.exe -C x86-64 --output-ji `cygpath -w /home/Tony/julia/usr/lib/julia/inference0.ji` -f coreimg.jl
Reading symbols from /home/Tony/julia/usr/bin/julia-debug.exe...done.
(gdb) r
Starting program: /home/Tony/julia/usr/bin/julia-debug.exe -C x86-64 --output-ji D:\\cygwin64\\home\\Tony\\julia\\usr\\lib\\julia\\inference0.ji -f coreimg.jl
[New Thread 8824.0x18bc]

Program received signal SIGSEGV, Segmentation fault.
0x00000001002c00b4 in julia.new_0 ()
(gdb) bt
#0  0x00000001002c00b4 in julia.new_0 ()
#1  0x000000006ea1dc8c in jl_apply (f=0x1012eec70, args=0xd7f078, nargs=2)
    at /home/Tony/julia/src/julia.h:1400
#2  0x000000006ea21e5b in jl_trampoline (F=0x1012eec70, args=0xd7f078, nargs=2)
    at /home/Tony/julia/src/builtins.c:1021
#3  0x000000006ea1038e in jl_apply (f=0x1012eec70, args=0xd7f078, nargs=2)
    at /home/Tony/julia/src/julia.h:1400
#4  0x000000006ea16ec6 in jl_apply_generic (F=0x1012eebf0, args=0xd7f078, nargs=2)
    at /home/Tony/julia/src/gf.c:1916
#5  0x000000006ea27f5d in jl_apply (f=0x1012eebf0, args=0xd7f078, nargs=2)
    at /home/Tony/julia/src/julia.h:1400
#6  0x000000006ea2841d in do_call (f=0x1012eebf0, args=0x1012f6448, nargs=2, eval0=0x0,
    locals=0x0, nl=0, ngensym=0) at /home/Tony/julia/src/interpreter.c:65
#7  0x000000006ea2908c in eval (e=0x1012eec30, locals=0x0, nl=0, ngensym=0)
    at /home/Tony/julia/src/interpreter.c:213
#8  0x000000006ea280b9 in jl_interpret_toplevel_expr (e=0x1012eec30)
    at /home/Tony/julia/src/interpreter.c:27
#9  0x000000006ea46aa9 in jl_toplevel_eval_flex (e=0x1012eec10, fast=1)
    at /home/Tony/julia/src/toplevel.c:525
#10 0x000000006ea46d6f in jl_parse_eval_all (fname=0x6f9a6e1d <system_image_path+1309> "boot.jl",
    len=8) at /home/Tony/julia/src/toplevel.c:575
#11 0x000000006ea46fbf in jl_load (fname=0x6f9a6e1d <system_image_path+1309> "boot.jl", len=8)
    at /home/Tony/julia/src/toplevel.c:615
#12 0x000000006ea332b5 in _julia_init (rel=JL_IMAGE_JULIA_HOME) at /home/Tony/julia/src/init.c:575
#13 0x000000006ea34a02 in julia_init (rel=JL_IMAGE_JULIA_HOME) at /home/Tony/julia/src/task.c:278
#14 0x0000000000402e0e in wmain (argc=1, argv=0x2ab0b00, envp=0x2ab3800)
    at /home/Tony/julia/ui/repl.c:605
#15 0x000000000040140c in __tmainCRTStartup ()
    at /usr/src/debug/mingw64-x86_64-runtime-4.0.2-1/crt/crtexe.c:329
#16 0x000000000040153b in mainCRTStartup ()
    at /usr/src/debug/mingw64-x86_64-runtime-4.0.2-1/crt/crtexe.c:212
```

**Comment by tkelman** [comment]
It looks as though the segfault was introduced by 2ab00adc4d7c2ae76c5a53000f357a8609d7e1ec. Prior to that it was giving broken modules, but at least starting bootstrap.

**Comment by vtjnash** [comment]
Thanks for that.  I have the fix, but had been confused about why it broke

**Pull request merged**